### ISSUE TITLE
[Feature] Support managedBy for integration with MultiKueue

### ISSUE THREAD

**Issue by mimowo** [ray-project/kuberay#2544]
### Search before asking

- [X] I had searched in the [issues](https://github.com/ray-project/kuberay/issues) and found no similar feature requirement.

### Description

The support for the managedBy field which can delegate reconciliation from built-in controller, to a custom one.

The semantics of the field are:

whenever the value is set, and it does not point to the built-in operator, then skip reconciliation
- the field is immutable
- the field is not defaulted

For mpi-operator and training-operator we add validation to only support: null, reserved value (like kubeflow.org/mpi-operator), and `kueue.x-k8s.io/multikueue`.

### Use case

Integration with [MultiKueue](https://kueue.sigs.k8s.io/docs/tasks/manage/setup_multikueue/) (multi-cluster Kueue):
- simpler installation (just follow the standard installation path), otherwise only installation of RayJob CRDs is required
- support for mixed setup in one cluster - some RayJobs could be run by MultiKueue and some by the default operator

### Related issues

The managedBy field is already adopted by:
- [k8s Job](https://github.com/kubernetes/enhancements/issues/4368) (beta in 1.32)
- [JobSet](https://github.com/kubernetes-sigs/jobset/issues/440) (released in 0.6)
- [MPI-operator](https://github.com/kubeflow/mpi-operator/issues/646) (released in 0.6)
- [training-operator](https://github.com/kubeflow/training-operator/issues/2193) (planned for release in Jan 2025)

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

**Comment by kevin85421** [comment]
Hi @mimowo, thank you for opening the issue! I’m not familiar with how MultiQueue works. Who creates the Ray Pods for the RayJob CR if it is managed by MultiQueue? Does MultiQueue also install a KubeRay operator?

**Comment by mimowo** [comment]
> I’m not familiar with how MultiQueue works

Sure, let me start with the basic design assumption (for more details I will refer you to the KEPs): [MultiKueue](https://github.com/kubernetes-sigs/kueue/tree/main/keps/693-multikueue) and [Job managed-by mechanism](https://github.com/kubernetes/enhancements/tree/master/keps/sig-apps/4368-support-managed-by-for-batch-jobs)

In MultiKueue we have 2 types of clusters: management cluster, and worker clusters (one to many). A user interacts only with the management cluster, but the pods and the actual execution is performed on the selected worker cluster. A copy of the job is created on that worker cluster by the MultiKueue controller. That copy of the job has `spec.managedBy` field set to null by the MultiKueue controller to delegate the actual execution on the worker cluster to the default controller. Once the copy is getting processed on the worker cluster the MultiKueue controller updates the status of the job on the management cluster (we call live update), so that the user interacting with the management cluster can see the progress.

To prevent creation of pods and any computations on the management cluster we introduce the `spec.managedBy` field - when the field is set to any custom value, the built-in job controller is "disabled" for that job. On top of that, to prevent complications when changing the "Job manager entity" we assume the field is immutable.

> Who creates the Ray Pods for the RayJob CR if it is managed by MultiQueue? 

On the "management" cluster (the cluster where `.spec.managedBy=kueue.x-k8s.io/multikueue` the Pods are not created at all, because the built-in controller is "disabled" (the new feature). On the worker cluster the pods are created, because the copy of the Job created by the user has "nulled" then `.spec.managedBy`.

> Does MultiQueue also install a KubeRay operator?

Nope, this is entirely to the cluster administrator. The admin will install KubeRay in the version which supports managedBy semantics on the management cluster.

**Comment by andrewsykim** [comment]
@mimowo does that mean every API supported by Kueue now has to add a `spec.managedBy` field? Why not use OwnerReference with a name mapping to Kueue or a label that every object already has?

**Comment by mimowo** [comment]
> @mimowo does that mean every API supported by Kueue now has to add a spec.managedBy field?

Kueue manages only RayCluster and RayJob, so these two APIs would need to have it, at least in the way we foresee the integration.  

>  Why not use OwnerReference with a name mapping to Kueue or a label that every object already has?

Maybe, I'm not sure exactly what you mean here. I suppose there might be an alternative mechanism, but since the implementation is very simple `if custom managedBy { return }`  (+ some webhooks to verify immutability) it seems acceptable. Let me know if I'm missing something,

**Comment by mimowo** [comment]
> @mimowo have we considered moving managedBy into ObjectMeta instead?

Yes, this was the original proposal in the k8s Job - if you look here https://github.com/kubernetes/enhancements/issues/4368 you see in the alpha phase it all started with "managed-by" label.
However, after discussions with k8s API reviewers (at length) we decided it is better to use a field, I have tried to capture the reasons [here](https://github.com/kubernetes/enhancements/tree/master/keps/sig-apps/4368-support-managed-by-for-batch-jobs#managed-by-label), the main being: "no clear indication if the label is supported, basically no fail-fast mechanism.".

**Comment by mimowo** [comment]
cc @mszadkow who is willing to work on the prototype

**Comment by mszadkow** [comment]
/assign
@andrewsykim thanks for the support readiness

**Comment by kevin85421** [comment]
> To prevent creation of pods and any computations on the management cluster we introduce the `spec.managedBy` field 

This seems to have the same behavior as `suspend: true`. Why not reuse the `suspend` field instead?

**Comment by mimowo** [comment]
> Do we need to install the KubeRay operator in the management cluster? The RayJob / RayCluster CRDs seem to be enough. In that case, Pods will not be created.

Technically we don't, but this makes the user experience worse because:
- installation is more complex, most users/admins just want to run `helm install kuberay-operator kuberay/kuberay-operator --version 1.3.0` and installing just CRDs makes the process susceptible to human error
- we want to support hybrid use of clusters (in one queue you can execute Jobs (RayJobs), in the other delegate to workers)

>  this seems to have the same behavior as suspend: true. Why not reuse the suspend field instead? Does managedBy prevent the KubeRay operator from receiving resource events for the RayJob CRs?

The reuse of the suspend field was discuss during the proposal for k8s Job. The want to enable users to get live-feedback about the progress the Job is doing in the worker cluster (and so update its status), there are in-house integrations like UI which monitor the progress. OTOH, some UI or integrations would not display the progress at all if the Job is suspended. One can also say it would be violation of the API contract to increase the `status.active` field while suspend=true (for the Job on the management cluster).

**Comment by kevin85421** [comment]
> we want to support hybrid use of clusters (in one queue you can execute Jobs (RayJobs), in the other delegate to workers)

makes sense

> The reuse of the suspend field was discuss during the proposal for k8s Job. The want to enable users to get live-feedback about the progress the Job is doing in the worker cluster (and so update its status). There are in-house integrations like UI which monitor the Job progress, which we can support OOTB this way. OTOH, some UI or integrations would not display the progress at all if the Job is suspended. One can also say it would be violation of the API contract to increase the status.active field while suspend=true (for the Job on the management cluster).

Thank you for the explanation. What do you mean by "progress" (e.g., status.succeeded)?

> some UI or integrations would not display the progress at all if the Job is suspended

out of curiosity, are there any UI open-source projects to display K8s Job progress?

**Comment by mimowo** [comment]
> What do you mean by "progress" (e.g., status.succeeded)?

Mostly succeeded, but any status field could be conveying something useful. For example, and admin seeing  high `.status.failed` can check early on, and delete the faulty job early.

> out of curiosity, are there any UI open-source projects to display K8s Job progress?

There are some, like [k9s](https://k9scli.io/) or [k8s dashboard](https://kubernetes.io/docs/tasks/access-application-cluster/web-ui-dashboard/), or [k8s lens](https://k8slens.dev/) (I'm probably missing some too). AFAIK the projects only mostly focus on the `status.succeeded` field, but could be extended relatively easily.

**Comment by kevin85421** [comment]
@mimowo thank you for the explanations! The request makes sense to me. Looking forward to the contributions!
